File: Assets/src/model/actors/Status.cs
using System;
using System.Collections.Generic;
using System.Linq;

/// Trying to reapply an existing status will call "Stack"
[Serializable]
public abstract class Status : IStepModifier {
  [field:NonSerialized] /// Controller only
  public event Action OnRemoved;
  private StatusList m_list;
  /// <summary>Should only be set by the StatusList.</summary>
  public StatusList list {
    get => m_list;
    set {
      if (value == null && m_list != null) {
        OnRemoved?.Invoke();
        End();
      }
      m_list = value;
      if (value != null) {
        Start();
      }
    }
  }

  public virtual bool isDebuff => false;
  public Actor actor => list?.actor;
  public virtual string displayName => Util.WithSpaces(GetType().Name.Replace("Status", ""));

  /// Called when list and actor are setup
  public virtual void Start() {}

  /// Called right before the status is removed. NOT called if the actor dies
  /// with the status on it!
  public virtual void End() {}

  public abstract string Info();

  /// The parameter will be of the same type as this type.
  /// Return true if this Status has consumed the other.
  public abstract bool Consume(Status other);

  public virtual void Step() {}

  object IModifier<object>.Modify(object input) {
    Step();
    return input;
  }

  /// <summary>Schedule this status for removal.</summary>
  public void Remove() {
    /// Must be scheduled because Remove() may be called within
    /// a Modifiers.Of() handler; modifying the list would
    /// cause a concurrent modification
    GameModel.main.EnqueueEvent(() => list?.Remove(this));
  }
}

/// Number stacking statuses will just "add" their numbers when re-applied rather than having the old one
/// removed
[Serializable]
public abstract class StackingStatus : Status {
  public virtual StackingMode stackingMode => StackingMode.Add;
  private int m_stacks;
  public virtual int stacks {
    get => m_stacks;
    set {
      m_stacks = value;
      if (value <= 0) {
        Remove();
      }
    }
  }

  public StackingStatus(int stacks) {
    this.stacks = stacks;
  }

  public StackingStatus() : this(1) {}

  public override bool Consume(Status otherParam) {
    var other = (StackingStatus) otherParam;
    switch (stackingMode) {
      case StackingMode.Add:
        this.stacks += other.stacks;
        return true;
      case StackingMode.Max:
        this.stacks = Math.Max(stacks, other.stacks);
        return true;
      case StackingMode.Ignore:
        return true;
      case StackingMode.Independent:
      default:
        return false;
    }
  }
}

public enum StackingMode { Add, Max, Ignore, Independent }

[Serializable]
public class StatusList {
  public Actor actor;
  public List<Status> list;

  public StatusList(Actor actor, List<Status> statuses) {
    this.actor = actor;
    this.list = statuses;
  }

  public StatusList(Actor actor) : this(actor, new List<Status>()) { }

  public void Add<T>(T status) where T : Status {
    var consumed = FindOfType<T>()?.Consume(status) ?? false;
    
    if (!consumed) {
      this.list.Add(status);
      status.list = this;
      OnStatusAdded(status);
    }
  }

  internal bool Has<T>() where T : Status {
    return this.list.Where((Status) => Status is T).Any();
  }

  public void RemoveOfType<T>() where T : Status {
    var toRemove = this.list.Where((status) => status is T).ToList();
    foreach (var status in toRemove) {
      Remove(status);
    }
  }

  public void Remove(Status status) {
    this.list.Remove(status);
    status.list = null;
    OnStatusRemoved(status);
  }

  internal T FindOfType<T>() where T : Status {
    return (T) (list.Find((s) => s is T));
  }

  private void OnStatusAdded(Status status) {
    foreach (var handler in actor.Of<IStatusAddedHandler>()) {
      handler.HandleStatusAdded(status);
    }
  }

  private void OnStatusRemoved(Status status) {
    foreach (var handler in actor.Of<IStatusRemovedHandler>()) {
      handler.HandleStatusRemoved(status);
    }
  }
}

File: web/src/model/Status.ts
import { STEP_MOD, type IStepModifier } from '../core/Modifiers';
import { GameModelRef } from './GameModelRef';
import type { StatusList } from './StatusList';
import type { Actor } from './Actor';

/**
 * Base class for all status effects.
 * Port of C# Status class from actors/Status.cs.
 *
 * Statuses implement IStepModifier so their Step() is called
 * during the modifier chain each turn.
 */
export abstract class Status implements IStepModifier {
  readonly [STEP_MOD] = true as const;

  private _list: StatusList | null = null;
  private removeScheduled = false;

  /** Called when removed (for controller/view hooks). */
  onRemoved: (() => void) | null = null;

  get list(): StatusList | null {
    return this._list;
  }

  /** Should only be set by StatusList. Setting to non-null calls Start(); setting to null calls End(). */
  set list(value: StatusList | null) {
    if (value == null && this._list != null) {
      this.onRemoved?.();
      this.End();
    }
    this._list = value;
    if (value != null) {
      this.Start();
    }
  }

  get actor(): Actor | null {
    return this._list?.actor ?? null;
  }

  get isDebuff(): boolean {
    return false;
  }

  /** Returns true if this status prevents the actor from moving. Used to suppress idle bob. */
  blocksMovement(): boolean {
    return false;
  }

  get displayName(): string {
    return this.constructor.name.replace(/Status$/, '').replace(/([A-Z])/g, ' $1').trim();
  }

  /** Called when list and actor are set up. */
  Start(): void {}

  /** Called right before the status is removed. NOT called if the actor dies with it. */
  End(): void {}

  /**
   * Return true if this Status has consumed the other (same type).
   * If consumed, the other status is discarded.
   */
  abstract Consume(other: Status): boolean;

  /** Called each turn via the step modifier chain. */
  Step(): void {}

  /** IStepModifier — step is called via modifier chain. */
  modify(input: object): object {
    this.Step();
    return input;
  }

  /** Schedule this status for removal. Safe to call during modifier processing. */
  Remove(): void {
    if (!this.removeScheduled) {
      this.removeScheduled = true;
      GameModelRef.main.enqueuEvent(() => this._list?.remove(this));
    }
  }

  /** Debuffs auto-remove on floor change. */
  handleFloorChanged(): void {
    if (this.isDebuff) {
      this.Remove();
    }
  }
}

// --- Stacking ---

export enum StackingMode {
  Add,
  Max,
  Ignore,
  Independent,
}

/**
 * Status that merges stacks when re-applied.
 * Port of C# StackingStatus.
 */
export abstract class StackingStatus extends Status {
  get stackingMode(): StackingMode {
    return StackingMode.Add;
  }

  private _stacks: number;

  get stacks(): number {
    return this._stacks;
  }

  set stacks(value: number) {
    this._stacks = value;
    if (value <= 0) {
      this.Remove();
    }
  }

  constructor(stacks = 1) {
    super();
    this._stacks = stacks;
  }

  Consume(other: Status): boolean {
    const otherStacking = other as StackingStatus;
    switch (this.stackingMode) {
      case StackingMode.Add:
        this.stacks += otherStacking.stacks;
        return true;
      case StackingMode.Max:
        this.stacks = Math.max(this.stacks, otherStacking.stacks);
        return true;
      case StackingMode.Ignore:
        return true;
      case StackingMode.Independent:
      default:
        return false;
    }
  }
}
